### ISSUE TITLE
sassc dependency breaks compatibility with tailwindcss-rails 

### ISSUE THREAD

**Issue by philipithomas** [thoughtbot/administrate#2091]
* What were you trying to do?

I'm trying to use [tailwind-rails v2.0.2](https://github.com/rails/tailwindcss-rails#conflict-with-sassc-rails) and importmaps. 

* What did you end up with (logs, or, even better, example apps are great!)?

I'm getting a `SassC::SyntaxError` when I run `rake assets:precompile`. Based on my debugging, it's because `tailwind-rails` is incompatible with the `sassc-rails` gem that administrate requires as a dependency:

> Tailwind uses modern CSS features that are not recognized by the sassc-rails extension that was included by default in the Gemfile for Rails 6. In order to avoid any errors like SassC::SyntaxError, you must remove that gem from your Gemfile.

* What versions are you running?

Rails 7.0
Administrate 0.16.0

**Comment by AnalyzePlatypus** [comment]
Please fix this, I have the same problem

**Comment by n-studio** [comment]
The solution is pretty simple: we need to get rid of all JS/CSS dependencies and just mention how to add them in the README.

We must let developers choose the way they handle their assets, whether it's using sassc, webpack or importmap.

Maybe we can add some generator `rails g administrate js:importmap` or something like that.

**Comment by etozzato** [comment]
`administrate` is a `Rails::Engine`, which means it is a Rails app. Would it be useful to refactor the code following the 'update to rails 7 guides' and homogenize the handling of resources?

**Comment by pablobm** [comment]
Unfortunately, this may not be simple at all. Before Rails 7, there were two main ways of handling assets: Webpacker and Sprockets. We have tried to make Administrate work with both, but it's difficult work and complex to reproduce issues.

Now Rails 7 has arrived with yet another way to do things. We'll have to spend the time to make it work there too, and we would rather that Administrate kept working with previous versions of Rails (at the moment we support Rails 5+).

I haven't used Rails 7 yet and I haven't really read much about how the new way to work with assets is. I will have to eventually, but right now I'm a bit short of time so I don't know when I'll get to sit down and figure this out :-/

Fortunately, it doesn't need to be us maintainers who do this work. One of you may be able to volunteer and offer a first stab at the solution. It doesn't need to be a complete PR: an initial prototype could help communicate what exactly needs to happen, and to start testing if there are issues with back compatibility.

Any takers? :-)

**Comment by n-studio** [comment]
@pablobm 
I have a fork compatible with Rails 7, I'll try to make it retrocompatible and submit a PR this weekend.
Are you ok with the idea of adding generators to add assets in the Rails app? This would be a breaking change but the upgrade would be easy and straightforward.

**Comment by n-studio** [comment]
@pablobm Another way to fix this issue is to compile the assets and commit the compiled assets. This would avoid breaking changes. Would you prefer this method?

**Comment by pablobm** [comment]
@n-studio - I would consider it. If you create a PR so that I can see what it looks like, it can help me understand.

**Comment by n-studio** [comment]
@pablobm Here is my PR https://github.com/thoughtbot/administrate/pull/2116
I hope you like that solution.

**Comment by dvruette** [comment]
I've been running into the same issue and the PR by @n-studio seems to fix the issue, thank you so much!

As a hotfix for anyone else running into this, I've added the branch with the fix to the Gemfile like so:
```ruby
gem "administrate", github: "n-studio/administrate", branch: "compile-assets"
```

**Comment by philipithomas** [comment]
@dvruette 

> I've been running into the same issue and the PR by @n-studio seems to fix it, thank you so much!
> 
> As a hotfix for anyone else running into this, I've added the branch with the fix to the Gemfile like so:
> 
> ```ruby
> gem "administrate", github: "n-studio/administrate", branch: "compile-assets"
> ```

This is a completely valid fix - but it will break if @n-studio deletes that particular branch in the future. And, it's a development branch - not `main` - so it's not reasonable to expect it to stick around. So, please fork the code to your own Github if you want to follow this approach! (and replace `github: "n-studio/administrate"` with your own fork)

**Comment by tkoenig** [comment]
What I did for some projects which still depend on Sass was [creating my own compressor](https://guides.rubyonrails.org/asset_pipeline.html#using-your-own-compressor), which skips the SassC compression if it fails.

```ruby
class SkippingSassCompressor
  def compress(string)
    options = { syntax: :scss, cache: false, read_cache: false, style: :compressed}
    begin
      Sprockets::Autoload::SassC::Engine.new(string, options).render
    rescue => e
      puts "Could not compress '#{string[0..65]}'...: #{e.message}, skipping compression"
      string
    end
  end
end
```

Since the skipped (tailwind) files, are not compressed this way I just added the `--minify` option to the `tailwindcss`CLI to compress them as well.

```json
"build:css": "tailwindcss -i ./app/assets/stylesheets/home.tailwind.css -o ./app/assets/builds/home.css --minify"
``` 

I did not test this especially with `administrate`, but have this working on multiple projects where we want to slowly migrate to tailwind.

**Comment by januszm** [comment]
Would this help here? https://github.com/sass/sassc-ruby/issues/220#issuecomment-1074825625

**Comment by supert56** [comment]
I ended up going down the same route as @tkoenig. It seems to work well for the time being 👍

**Comment by jankeesvw** [comment]
We are currently using the fork by @n-studio, but we would love to run the official version with the new features. Is anyone working on this still? We are using the new Rails 7 defaults (cssbundling-rails & jsbundling-rails).

**Comment by czepesch** [comment]
unfortunately @n-studio's fork breaks Stimulus..

**Comment by jusko** [comment]
This was quite a time consuming issue to get to the bottom of. Thanks @tkoenig and @n-studio for offering stop gap solutions.

**Comment by geniuskidkanyi** [comment]
Moving to dartsass-rails will fix this issue, i currently got it working with dartsass, i had to generate the administrate gem stylesheets in my project
here is a link to my fork with dart-sass 
[fork works with tailwind](https://github.com/geniuskidkanyi/administrate/tree/asset-fix)

**Comment by geniuskidkanyi** [comment]
> Moving to dartsass-rails will fix this issue, i currently got it working with dartsass, i had to generate the administrate gem stylesheets in my project here is a link to my fork with dart-sass [fork works with tailwind](https://github.com/geniuskidkanyi/administrate/tree/asset-fix)

i just updated the branch, now you dont need to generate the assets in your project

**Comment by oroth8** [comment]
I also tried various forks and solutions mentioned on this issue page. Nothing worked. I created a fork myself https://github.com/oroth8/administrate and it works perfectly with my rails 7 tailwind configured app.

**Comment by jankeesvw** [comment]
I would love it if there was official support for this. 
Is there anyway we can get this merged (maybe as a major version bump)?

**Comment by sedubois** [comment]
Actually, it looks like we managed to continue to use administrate, even with its dependency on the old sassc-rails, while the rest of our app changes to cssbundling-rails with Dart Sass (for our legacy views) and Tailwind (for new views). The key for us was to ensure `config.assets.css_compressor` was set to `nil` for all Rails environments. Without this we encountered errors such as `ActionView::Template::Error: Error: Function rgb is missing argument $green` and I think also errors of the type `ActionView::Template::Error: Asset X was not declared to be precompiled in production`. So far we haven't needed to use one of the administrate forks suggested in this PR. Also see https://github.com/rails/cssbundling-rails/pull/129.

**Comment by kaka-ruto** [comment]
Thank you all for your solutions to this, it's been a headache! You solution @sedubois worked for me, thank you for sharing it. I just dropped `Rails.application.config.assets.css_compressor = nil` in `config/initializers/assets.rb` and it worked.

**Comment by marcobeffa** [comment]
Thank you all for your solutions to this, it's been a headache! You solution @kaka-ruto @sedubois Thank You.  I just dropped Rails.application.config.assets.css_compressor = nil in config/initializers/assets.rb and it worked.

**Comment by nickcharlton** [comment]
I'm going to close this we've just merged in #2397, which will mean that in future release we'll be bundling assets in the Administrate Gem and so we've been able to drop any dependencies on assets from those using the gem.

I'm expecting this will mean Tailwind will work fine now, regardless of how you go about bundling the assets in your applications. 

I'm going to start releasing some pre-release versions of the gem in the coming days, as I'm sure I'm about to break stuff for people, and so we can get feedback and make sure there's nice migration help sorted out.
